gpudrive/integrations/puffer/ppo.py

```diff
@@ -643,27 +747,103 @@ def sort_training_data(self):
         self.step = 0
         return idxs
 
-    def flatten_batch(self, advantages_np):
-        advantages = torch.from_numpy(advantages_np).to(self.device)
-        b_idxs, b_flat = self.b_idxs, self.b_idxs_flat
-        self.b_actions = self.actions.to(self.device, non_blocking=True)
-        self.b_logprobs = self.logprobs.to(self.device, non_blocking=True)
-        self.b_dones = self.dones.to(self.device, non_blocking=True)
-        self.b_values = self.values.to(self.device, non_blocking=True)
-        self.b_advantages = (
-            advantages.reshape(
-                self.minibatch_rows, self.num_minibatches, self.bptt_horizon
+    def flatten_batch(self, advantages_np, filter_mask=None):
+        """Prepare the batch of transitions for model updating."""    
+        
+        if filter_mask is not None:
+
+            # Get the indices of transitions to keep 
+            kept_indices = np.nonzero(filter_mask)[0]
+            total_kept = len(kept_indices)
+                    
+            # Determine how many transitions per minibatch (floor division)
+            transitions_per_mb = total_kept // self.num_minibatches
+            
+            # We need at least one transition per minibatch
+            if transitions_per_mb < 32:
+                transitions_per_mb = 64
+                
+                warnings.warn(f"Low adv. filtering retention rate: Only kept {len(kept_indices)} / {len(advantages_np)} transitions ({transitions_per_mb} per minibatch) \n Consider adjusting the advantage threshold factor or increase the batch_size.", UserWarning) 
+                
+                # If we don't have enough, sample with replacement
+                if total_kept < self.num_minibatches:
+                    kept_indices = np.random.choice(kept_indices, self.num_minibatches, replace=True)
+                    total_kept = len(kept_indices)
+            
+            # Calculate total transitions to use (divisible by num_minibatches)
+            transitions_to_use = transitions_per_mb * self.num_minibatches
+            
+            np.random.shuffle(kept_indices)
+            kept_indices = kept_indices[:transitions_to_use]
```

Comment:
What is this line supposed to do?

Reply:
It ensures that the number of transitions is divisible across the # of mini batches